src/WebJobs.Script/Description/CSharp/CSharpFunctionInvoker.cs

```diff
@@ -0,0 +1,373 @@
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.Azure.WebJobs.Host;
+using Microsoft.Azure.WebJobs.Script.Binding;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Scripting;
+using Microsoft.CodeAnalysis.Scripting;
+
+namespace Microsoft.Azure.WebJobs.Script.Description
+{
+    public class CSharpFunctionInvoker : ScriptFunctionInvokerBase
+    {
+        private const string ScriptClassName = "Submission#0";
+
+        private readonly ScriptHost _host;
+        private readonly DictionaryJsonConverter _dictionaryJsonConverter = new DictionaryJsonConverter();
+        private readonly TraceWriter _fileTraceWriter;
+        private readonly FunctionMetadata _functionMetadata;
+        private readonly BindingMetadata _trigger;
+        private readonly Collection<FunctionBinding> _inputBindings;
+        private readonly Collection<FunctionBinding> _outputBindings;
+        private readonly FileSystemWatcher _fileWatcher;
+        private readonly string _triggerParameterName;
+        private readonly bool _omitInputParameter;
+        private readonly IFunctionEntryPointResolver _functionEntryPointResolver;
+        private MethodInfo _function;
+
+        public CSharpFunctionInvoker(ScriptHost host, BindingMetadata trigger, FunctionMetadata metadata, bool omitInputParameter,
+            Collection<FunctionBinding> inputBindings, Collection<FunctionBinding> outputBindings, IFunctionEntryPointResolver functionEntryPointResolver)
+        {
+            _host = host;
+            _trigger = trigger;
+            _triggerParameterName = trigger.Name;
+            _omitInputParameter = omitInputParameter;
+            _inputBindings = inputBindings;
+            _outputBindings = outputBindings;
+            _functionMetadata = metadata;
+            _functionEntryPointResolver = functionEntryPointResolver;
+            _fileWatcher = InitializeFileWatcher(host);
+
+            if (_host.ScriptConfig.FileLoggingEnabled)
+            {
+                string logFilePath = Path.Combine(_host.ScriptConfig.RootLogPath, "Function", _functionMetadata.Name);
+                _fileTraceWriter = new FileTraceWriter(logFilePath, TraceLevel.Verbose);
+            }
+            else
+            {
+                _fileTraceWriter = NullTraceWriter.Instance;
+            }
+        }
+
+        public static string[] DefaultAssemblyReferences
+            => new string[]
+            {
+                "System",
+                "System.Core",
+                "System.Xml",
+                "System.Net.Http",
+                "Microsoft.WindowsAzure.Storage"
+            };
+
+        public static Assembly[] DefaultAssemblies
+            => new Assembly[]
+            {
+                typeof(TraceWriter).Assembly,
+                typeof(object).Assembly,
+                typeof(TimerInfo).Assembly
+            };
```

Comment:
so we're going to be using the project.json syntax. I was just worried that we would need to invent our own package restore and csproj like system to support external dependencies, class libraries, multi-file functions, etc. I played with roslyn auto compilation before and had to pass in an array of reference assemblies and an array of source files so it wasn't apparent how we would achieve that without inventing our own build conventions which I thought we should stay away from. but if there is a clear plan to use the project.json semantics then I'm fine with that.